cpp/src/scalar/scalar.cpp

```diff
@@ -56,4 +95,339 @@ std::string string_scalar::to_string(rmm::cuda_stream_view stream) const
   return result;
 }
 
+template <typename T>
+fixed_point_scalar<T>::fixed_point_scalar() : scalar(data_type(type_to_id<T>())){};
+
+template <typename T>
+fixed_point_scalar<T>::fixed_point_scalar(rep_type value,
+                                          numeric::scale_type scale,
+                                          bool is_valid,
+                                          rmm::cuda_stream_view stream,
+                                          rmm::mr::device_memory_resource* mr)
+  : scalar{data_type{type_to_id<T>(), static_cast<int32_t>(scale)}, is_valid, stream, mr},
+    _data{value}
+{
+}
+
+template <typename T>
+fixed_point_scalar<T>::fixed_point_scalar(rep_type value,
+                                          bool is_valid,
+                                          rmm::cuda_stream_view stream,
+                                          rmm::mr::device_memory_resource* mr)
+  : scalar{data_type{type_to_id<T>(), 0}, is_valid, stream, mr}, _data{value}
+{
+}
+
+template <typename T>
+fixed_point_scalar<T>::fixed_point_scalar(T value,
+                                          bool is_valid,
+                                          rmm::cuda_stream_view stream,
+                                          rmm::mr::device_memory_resource* mr)
+  : scalar{data_type{type_to_id<T>(), value.scale()}, is_valid, stream, mr}, _data{value.value()}
+{
+}
+
+template <typename T>
+fixed_point_scalar<T>::fixed_point_scalar(rmm::device_scalar<rep_type>&& data,
+                                          numeric::scale_type scale,
+                                          bool is_valid,
+                                          rmm::cuda_stream_view stream,
+                                          rmm::mr::device_memory_resource* mr)
+  : scalar{data_type{type_to_id<T>(), scale}, is_valid, stream, mr},
+    _data{std::forward<rmm::device_scalar<rep_type>>(data)}
+{
+}
+
+template <typename T>
+typename fixed_point_scalar<T>::rep_type fixed_point_scalar<T>::value(
+  rmm::cuda_stream_view stream) const
+{
+  return _data.value(stream);
+}
+
+template <typename T>
+T fixed_point_scalar<T>::fixed_point_value(rmm::cuda_stream_view stream) const
+{
+  return value_type{
+    numeric::scaled_integer<rep_type>{_data.value(stream), numeric::scale_type{type().scale()}}};
+}
+
+template <typename T>
+typename fixed_point_scalar<T>::rep_type* fixed_point_scalar<T>::data()
+{
+  return _data.data();
+}
+
+template <typename T>
+typename fixed_point_scalar<T>::rep_type const* fixed_point_scalar<T>::data() const
+{
+  return _data.data();
+}
+
+template class fixed_point_scalar<numeric::decimal32>;
```

Comment:
Comment here like the ones below saying "These are the valid types of fixed-point scalars"?